base_layer/wallet/src/storage/sqlite_db/wallet.rs

```diff
@@ -438,92 +442,136 @@ impl WalletBackend for WalletSqliteDatabase {
-        );
-    }
+/// Derive a secondary database key
+fn derive_secondary_key(
+    passphrase: &SafePassword,
+    params: Argon2Parameters,
+    salt: &String,
+) -> Result<WalletSecondaryEncryptionKey, WalletStorageError> {
+    // Derive a secondary key from the existing passphrase and salt
+    let mut secondary_key = WalletSecondaryEncryptionKey::from(SafeArray::default());
+    argon2::Argon2::new(params.algorithm, params.version, params.params)
+        .hash_password_into(passphrase.reveal(), salt.as_bytes(), secondary_key.reveal_mut())
+        .map_err(|e| WalletStorageError::AeadError(e.to_string()))?;
+
+    Ok(secondary_key)
+}
+
+/// Encrypt the main database key using the secondary key
+fn encrypt_main_key(
+    secondary_key: &WalletSecondaryEncryptionKey,
+    main_key: &WalletMainEncryptionKey,
+    version: u8,
+) -> Result<Vec<u8>, WalletStorageError> {
+    // Set up the authenticated data
+    let mut aad = MAIN_KEY_AAD_PREFIX.as_bytes().to_owned();
+    aad.push(version);
+
+    // Encrypt the main key
+    let cipher = XChaCha20Poly1305::new(Key::from_slice(secondary_key.reveal()));
+    let encrypted_main_key = encrypt_bytes_integral_nonce(&cipher, aad, Hidden::hide(main_key.reveal().clone()))
+        .map_err(WalletStorageError::AeadError)?;
+
+    Ok(encrypted_main_key)
+}
 
-    Ok(cipher)
+/// Decrypt the main database key using the secondary key
+fn decrypt_main_key(
+    secondary_key: &WalletSecondaryEncryptionKey,
+    encrypted_main_key: &str,
+    version: u8,
+) -> Result<WalletMainEncryptionKey, WalletStorageError> {
+    // Set up the authenticated data
+    let mut aad = MAIN_KEY_AAD_PREFIX.as_bytes().to_owned();
+    aad.push(version);
+
+    // Authenticate and decrypt the main key
+    let cipher = XChaCha20Poly1305::new(Key::from_slice(secondary_key.reveal()));
+
+    Ok(WalletMainEncryptionKey::from(
+        decrypt_bytes_integral_nonce(
+            &cipher,
+            aad,
+            &from_hex(encrypted_main_key).map_err(|e| WalletStorageError::ConversionError(e.to_string()))?,
+        )
+        .map_err(|_| WalletStorageError::InvalidPassphrase)?,
+    ))
 }
 
-fn get_cipher_for_db_encryption(
-    passphrase: SafePassword,
-    secondary_key_version: Option<String>,
-    secondary_key_salt: Option<String>,
-    encrypted_main_key: Option<String>,
-    secret_seed: &Option<Hidden<String>>,
-    conn: &SqliteConnection,
+/// Prepare the database encryption cipher
+fn get_db_cipher(
+    database_connection: &WalletDbConnection,
+    passphrase: &SafePassword,
 ) -> Result<XChaCha20Poly1305, WalletStorageError> {
-    // We'll bind the version identifier to this domain before it's used
-    let main_key_domain = b"wallet_main_key_encryption_v".to_vec();
+    let conn = database_connection.get_pooled_connection()?;
+
+    // Fetch the database fields used for encryption, if they exist
+    let secondary_key_version = WalletSettingSql::get(&DbKey::SecondaryKeyVersion, &conn)?;
```

Comment:
Should technically also be in  a transaction since it could change half way through reading

Reply:
Would using a transaction catch this edge case? Even if not, I don't think it's a practical problem here. In the event that the fields were to be only partially updated when the reads happen, the subsequent key derivation will fail and return an error. The user can then simply try again to load the wallet.

Reply:
Will address in [this issue](https://github.com/tari-project/tari/issues/5177).